### DISCUSSION TITLE
Breaking changes not documented with 2.6

### DISCUSSION THREAD

**Discussion by mabdullahsari** [livewire/livewire#3740] [7. Bugs (archived)] [answered]
Unless these are regressions, I think that 2.6 should have been tagged as 3.0. 

**I'm very much grateful for this awesome package** and very lucky to have an extensive test suite to catch these off the bat.

Things I've encountered:

- `initializeX` methods were allowed to be `protected`, but now they cannot be found apparently requiring the access modifier to be `public` in order to work again.

- Using dynamically created & assigned properties in conjunction with Livewire's own dynamic property system with `getXProperty` no longer worked for me, the way I initially implemented it. The reason I did this, was to prevent N+1 on the initial mount of the component and dynamically retrieve an entry from the database for subsequent requests (because I did not want to pass along all of the model's attributes to the client).

Works in 2.5:

```php
public function mount(CartItem $item)
{
    $this->identifier = $item->id;
    $this->item = $item; // 👈 dynamically assigned property, there is no "item" class property defined
}

public function getItemProperty(): CartItem
{
    return Cart::find($this->identifier);
}
 ```
 
 Breaks in 2.6:
 
 ```bash
 Livewire component's [app.http.cart.livewire.item] public property [item] must be of type: [numeric, string, array, null, or boolean].
Only protected or private properties can be set as other types because JavaScript doesn't need to access them.
```

The fix I did was creating a private `$_item` property and referencing that in the `getItemProperty` method:

```php
public function getItemProperty(): CartItem
{
    return $this->_item ?? Cart::find($this->identifier);
}
```
Kind of defeats my clever trick, but it works again at least.

**Comment by jacovdbergh** [comment]
Found another breaking change in 2.6

In 2.5 the Livewire/Component __invoke looked like this:
```php
public function __invoke(Container $container, Route $route)
```

In 2.6 the $route parameter was removed:
```php
public function __invoke(Container $container)
```

We override the __invoke method to get route model binding working outside Livewire (for use in a middleware).

**Reply by calebporzio** [reply]
Mmmm I feared that was an issue. Check out this comment: https://github.com/livewire/livewire/pull/3495#issuecomment-912551555

Can you post your __invoke override code?

**Reply by calebporzio** [reply]
We can add `Route $route` back as a param to not break the signature, and just not use `$route` - but I'm curious if you are actually relying on the ability to change `$route`.

**Reply by calebporzio** [reply]
Ok, I added `Route $route` back, but `$route` won't be used internally anymore.

**Reply by georgeboot** [reply]
Awesome!

Shouldn't you have anyway injected the Request and take the route from there, instead of injecting the route directly? Anyways, something for v3 haha

**Comment by joshhanley** [comment, marked as answer]
@mabdullahsari thanks for reporting! They're aren't intentional breaking changes, so something must have caused a regression.

I'm not sure what would have caused dynamic properties to stop working, as we do support them (non of our tests have failed).

I will see if I can find out what PR's causing it in the morning 🙂

**Reply by mabdullahsari** [reply]
Hi @joshhanley 

I realize my way of using PHP's native dynamic properties in conjunction with Livewire's is esoteric, but to further elaborate on the issue:

I have an overview of cart items which are rendered via a foreach loop, but each rendered cart item is a Livewire component in itself. But because I already have these items at my disposal, it'd be wasteful to query the database again in each individual Livewire component. So, I'm passing the cart items to the individual Livewire components via mount:

```blade
@forelse ($items as $item)
    @livewire('b2b::cart:item', compact('item'), key($item->id))
@empty
    <x-empty-view />
@endforelse
```

and assigning them to a class property, **dynamically**:

```php
class Item extends Component
{
    public int $identifier;

    // There is no item property defined, however PHP allows this

    public function mount(CartItem $item, bool $editable = true)
    {
        $this->item = $item;
    }

    public function getItemProperty(): CartItem
    {
        return Cart::find($this->identifier);
    }
```

This makes use of PHP's native feature discussed in this RFC: https://wiki.php.net/rfc/deprecate_dynamic_properties
(I'm aware it's being discussed for deprecation, but for now it's still usable)

Livewire's dynamic properties only get invoked if the class does not contain such a property, but since it does for the **initial render**, I'm able to bypass a query and don't have to include conditionals in my corresponding blade view.

If the user decides to change something in **one** of the Livewire components, only that component will get updated and since it's not an initial render, the dynamic property will get invoked due to the `item` class property missing (which was set in the mount method for the initial render).

I hope my explanation is not too unwieldy to grasp what's going on. Just making clever use of when `__get` is invoked and when not.

**Reply by joshhanley** [reply]
@mabdullahsari thanks for explaining 🙂 

One area I just want to confirm though, as I may be missing something, but when I use nested components and say pass `$item` into the nested component, I don't get any extra queries run.

Are you saying that if you pass `$item` into your nested component it is executing an extra query? If so, something unusual is happening then.

I have a similar situation, where the cart and all items and details are loaded in the parent component, and the individual items are then passed down into the nested components. I do this, so I can run `$cart->load(['items'])` in the parent, which means I only have two queries running (hence no n+1 problem).

But I have a public property called `$item` in my nested component, which Livewire will use `$item` passed from the parent on the initial request and auto populate it for you from the database on any subsequent requests (provided `$item` is an eloquent model).

Can you elaborate on whether that process would work for you and why/ why not?

Thanks!

**Reply by mabdullahsari** [reply]
Hey @joshhanley 

It does not execute any extra queries during the initial render which is exactly what I want.

`$item` is not an Eloquent model, unfortunately. Also, it contains lots of attributes. That's why I decided to only pass along the id to the client as only a very small portion is being used in the view itself.

This was a nice way of bridging that gap, but no longer works 🤷🏽‍♂️ Would be nice if this behavior could get restored.

**Reply by mabdullahsari** [reply]
Some change is probably causing dynamically added props to get picked up as well now. I've checked the `wire:initial-data` attribute in the HTML response and the dynamically added property is definitely not there (in 2.5).

**Reply by mabdullahsari** [reply]
Oh, and I'm totally aware that my problem can be fixed using the new `Wirable` contract. However, this is still a breaking change that ideally should be fixed. Unfortunately I can't upgrade every single client project we have right away.

I hope the problem is clear to you? 🙌 Thanks again.

**Reply by joshhanley** [reply]
@mabdullahsari ah ok, not being an eloquent model makes sense.

So I believe the reason that isn't working now is due to this PR #3640.

Have a look at Calebs comment on that PR https://github.com/livewire/livewire/pull/3640#issuecomment-921298407

The preference would be not to revert this change as it's more of a bug fix. But at the same time don't want to cause you major problems. So would like your thoughts about that PR and Caleb's comment and the impact on you.

Is this something you use across a lot of your apps?

I'm open to ideas of how to solve it other than just reverting the PR.

Maybe we can find another way around this, like adding a config option or something, so you can set it back to the old way of operating, to give you time to make the changes, then in V3 we can remove the config option all together?

Or maybe there is another solution?

These are only my suggestions and Caleb may have other ideas, but hopefully we can find something workable.

**Reply by mabdullahsari** [reply]
I read Caleb's comment and it makes sense I guess. I'll change them one by one, not a huge deal. This is when an extensive test suite shines 🙂 However, you should really announce this change somewhere. You can't just expect people to understand without a proper public explanation. Caleb's comment definitely makes sense, but we had to dig through a merged PR for that. 👀

**Reply by joshhanley** [reply]
@mabdullahsari yeah can't bet a good test suite! 😁

Yep fair call, will have a look into where is best to document that.

**Comment by calebporzio** [comment]
@mabdullahsari - thanks for reporting. I've reverted the `initializeX` protected/public change
